File: python/heap-priority-queue/kth-largest-element-in-a-stream.py
import heapq

class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self.heap = []
        self.size = k
        for i in nums:
            heapq.heappush(self.heap, i)

        for i in range(len(nums) - k):
            heapq.heappop(self.heap)

    def add(self, val: int) -> int:
        heapq.heappush(self.heap, val)
        if len(self.heap) > self.size:
            heapq.heappop(self.heap)
        kth = heapq.heappop(self.heap)
        heapq.heappush(self.heap, kth)
        return kth

# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: c++/heap-priority-queue/kth-largest-element-in-a-stream.cc
class KthLargest {
public:
    KthLargest(int k, vector<int>& nums) {
        capacity_ = k;
        for (int num : nums) {
            pq_.push(-num);
        }

        while (pq_.size() > k) {
            pq_.pop();
        }
    }
    
    int add(int val) {
        pq_.push(-val);
        if (pq_.size() > capacity_) {
            pq_.pop();
        }

        return -pq_.top();
    }

private:
    int capacity_;
    std::priority_queue<int> pq_;
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */
 